src/targets/gpu/mlir_conv.cpp

```diff
@@ -0,0 +1,264 @@
+#include <migraphx/gpu/mlir_conv.hpp>
+#include <migraphx/manage_ptr.hpp>
+#include <migraphx/instruction.hpp>
+#include <migraphx/make_op.hpp>
+
+#include <migraphx/op/convolution.hpp>
+
+#include <migraphx/gpu/context.hpp>
+#include <migraphx/gpu/convolution.hpp>
+#include <migraphx/iterator_for.hpp>
+#include <migraphx/program.hpp>
+
+#include <migraphx/ranges.hpp>
+#include <migraphx/make_op.hpp>
+#include <migraphx/generate.hpp>
+#include <migraphx/program.hpp>
+#include <migraphx/gpu/kernel.hpp>
+#include <migraphx/gpu/target.hpp>
+#include <migraphx/gpu/hip.hpp>
+#include <migraphx/gpu/compile_hip.hpp>
+
+#include <utility>
+#include <functional>
+#include <algorithm>
+
+#ifdef    MIGRAPHX_MLIR_MIOPEN_SUPPORT
+#include <Miir.h>
+#endif // MIGRAPHX_MLIR_MIOPEN_SUPPORT
+
+#include <stdio.h>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+namespace gpu {
+
+struct mlir_apply
+{
+    module* prog                 = nullptr;
+    const mlir_conv* pass        = nullptr;
+
+    const char *mlir_kernel_name = "migraphx_conv2d";
+
+    std::unordered_map<uint64_t, instruction_ref> literal_map{};
+
+    struct execution_spec {
+      migraphx::value::binary binary;
+      size_t global_size;
+      size_t local_size;
+    };
+    std::unordered_map<std::string, execution_spec*> binary_map{};
+
+    context& get_context() const
+    {
+        assert(pass != nullptr);
+        assert(pass->ctx != nullptr);
+        return *pass->ctx;
+    }
+
+    void init()
+    {
+        assert(prog != nullptr);
+        assert(pass != nullptr);
+    }
+
+    execution_spec *make_mlir_binary(instruction_ref op_r)
```

Comment:
I think this should just return by value. We know if it didn't create the binary if `binary.empty()` is true.

Reply:
That would induce a copy, would be very expensive with the binary vector copy.

Reply:
NRVO should elide the copy, but even if that is not used, the compiler should implicitly move the value in C++14(according to [1579](http://open-std.org/JTC1/SC22/WG21/docs/cwg_defects.html#1579)), which will avoid a copy, and just do some pointer swaps.

Reply:
But the execution_spec is potentially shared from the unordered_map (which owns the object). That way multiple conv2d's with the same config will not call the compiler every time.

Reply:
I see, so if the map owns it, it should be `std::unordered_map<std::string, execution_spec>` and then this function can return a pointer to the element in the map(it doesn't seem like shared ownership is necessary), but you will need to add elements to the map with `binary_map.emplace()` so it can move(rather than copy) the object into the map.

Reply:
True. will do.

Reply:
Will use shared_ptr instead since we still want nullptr's in the map.

Reply:
Why do we want nullpointers in the map? Whats the difference between the key not being in the map and the key being in the map but the value is null?

Reply:
Wait, I see, null means it tried but failed to compile a kernel, whereas a missing key means it hasn't even tried.